File: solve.php
<?php

function solve()
{
    if (strcmp($GLOBALS['osets'][0]->getHash(), $GLOBALS['sol']->getHash()) == 0)
        die ("Puzzle is already solved.\n");

    $time = time();
    $iterations = 0;
    $nodes = 1;
    $GLOBALS['checked'][] = $GLOBALS['osets'][0]->getHash();
    while (1) {
        //get the chepest openset
        $iterations++;
        unset($cheapest);
        foreach ($GLOBALS['osets'] as $node) {
            if (!isset($cheapest))
                $cheapest = $node;
            if ($GLOBALS['algo'] == ASTAR) {
                if (isset($cheapest) && $node->getFofX() <= $cheapest->getFofX())
                    $cheapest = $node;
            } else if ($GLOBALS['algo'] == GREEDY) {
                if (isset($cheapest) && $node->getCost() < $cheapest->getCost())
                    $cheapest = $node;
            } else if ($GLOBALS['algo'] == DEPTHFIRST) {
                $cheapest = end($GLOBALS['osets']);
            } else //breadthfirst
                break;
        }
        if ($cheapest == NULL)
            die ("No open set found\n");

       /* if ($GLOBALS['verb'] == 1){
            echo "chose:\n ".$cheapest."\nfrom:\n";
            foreach ($GLOBALS['osets'] as $node) {
                echo $node."\n";
            }
            echo "\n";
        }*/
        //Get new moves from cheapest node, compare to solution and add to open sets
        $newnodes = $cheapest->genMoves();
        //print_r($cheapest->genMoves());
        //print_r($newnodes);
        foreach ($newnodes as $node) {
            if (strcmp($node->getHash(), $GLOBALS['sol']->getHash()) === 0) {
                echo "Solution Found:\n";
                echo $GLOBALS['csets'][0] . "\n";
                printSolution($node);
                echo "\nSolution found in " . (time() - $time) . " second(s) after " . $iterations . " iteration(s).\n";
                echo "{$nodes} nodes were generated in order to find the " . $node->getDist() . " moves of the solution.\n";
                echo "open list: " . sizeof($GLOBALS['osets']) . " closed set " . sizeof($GLOBALS['csets']) . " \n";
                die("\n");
            }
        }

        foreach ($newnodes as $new) {
            $GLOBALS['osets'][] = $new;
            $GLOBALS['checked'][] = $new->getHash();
            $nodes++;
        }

        //Add cheapest to closed sets and remove from open sets
        $GLOBALS['csets'][] = $cheapest;
        $index = -1;
        while (isset($GLOBALS['osets'][++$index]))
            if ($cheapest->getId() === $GLOBALS['osets'][$index]->getId())
                array_splice($GLOBALS['osets'], $index, 1);


        if ($GLOBALS['verb'] == 1)
            //echo $cheapest . "\n";
            ;
        else if ($iterations % 100 == 0)
            echo ".";
    }
}

function printSolution($endNode)
{
    $pid = $endNode->getParent();
    foreach ($GLOBALS['osets'] as $node) {
        $tid = $node->getId();
        if ($tid != NULL && $tid == $pid)
            printSolution($node);

    }
    foreach ($GLOBALS['csets'] as $node) {
        $tid = $node->getId();
        if ($tid != NULL && $tid == $pid)
            printSolution($node);
    }
    echo $endNode . "\n";
}

?>

File: solve.cpp
//
// Created by Robert JONES on 2016/11/16.
//

#include "npuzzle.h"

void printSolution(Node *endNode) {
    int pid = endNode->getParent();
    for (vector<Node>::iterator node = g_osets.begin(); node < g_osets.end(); node++) {
        int tid = node->getId();
        if (tid == pid)
            printSolution(&(*node));
    }
    for (vector<Node>::iterator node = g_csets.begin(); node < g_csets.end(); node++) {
        int tid = node->getId();
        if (tid == pid)
            printSolution(&(*node));
    }
    cout << endNode->toString();
}

void solve() {
    if (strcmp(g_osets.at(0).getHash().c_str(), g_sol->getHash().c_str()) == 0) {
        cout << "Puzzle is already solved.\n";
        exit(EXIT_SUCCESS);
    }

    //time = time();
    int iterations = 0;
    int nodes = 1;
    g_checked.push_back(g_osets.at(0).getHash());

    cout << "oset size: " << g_osets.size() << endl;

    while (1) {
        //get the chepest openset
        iterations++;
        Node *cheapest = NULL;
        for (vector<Node>::iterator node = g_osets.begin(); node < g_osets.end(); node++) {
            if (cheapest == NULL) {
                cheapest = &(*node);
            } else {
                if (g_algo == ASTAR) {
                    if (cheapest != NULL && node->getFofX() <= cheapest->getFofX()) {
                        cheapest = &(*node);
                    }
                } else if (g_algo == GREEDY) {
                    if (cheapest != NULL && node->getCost() < cheapest->getCost())
                        cheapest = &(*node);
                } else if (g_algo == DEPTHFIRST) {
                    cheapest = &(*g_osets.end());
                } else //breadthfirst
                    break;
            }
        }
        if (cheapest == NULL) {
            cout << "No open set found\n";
            exit(EXIT_SUCCESS);
        }
        //Get new moves from cheapest node, compare to solution and add to open sets
        vector<Node> newnodes = cheapest->genMoves();
        for (vector<Node>::iterator node = newnodes.begin(); node < newnodes.end(); node++) {
            if (strcmp(node->getHash().c_str(), g_sol->getHash().c_str()) == 0) {
                cout << "\n\nSolution Found:\n";
                cout << node->toString() << "\n";
                cout << "steps to solution:\n";
                printSolution(&(*node));
                //cout << "\nSolution found in " << (time() - time) << " second(s) after "<< iterations << " iteration(s).\n";
                cout << "{nodes} nodes were generated in order to find the " << node->getDist()
                     << " moves of the solution.\n";
                cout << "open list: " << g_osets.size() << " closed set " << g_csets.size() << " \n";
                cout << "\n";
                exit(EXIT_SUCCESS);
            }
        }


        for (vector<Node>::iterator newNode = newnodes.begin(); newNode < newnodes.end(); newNode++) {
            g_osets.push_back((*newNode));
            g_checked.push_back(newNode->getHash());
        }

        //Add cheapest to closed sets and remove from open sets
        g_csets.push_back(*cheapest);
        int index = -1;
        int max = g_osets.size();
        while (++index < max) {
            if (cheapest->getId() == g_osets.at(index).getId()) {
                g_osets.erase(g_osets.begin() + index);
                break;
            }
        }
        if (g_verb == 1)
            cout << cheapest->toString() << "\n";
        else if (iterations % 100 == 0)
            cout << ".";
    }
}

